presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -0,0 +1,542 @@
+        }
+        catch (Exception e) {
+            throw new PrestoException(COMPILER_ERROR, e.getCause());
+        }
+
+        return () -> {
+            try {
+                return projectionClass.newInstance();
+            }
+            catch (ReflectiveOperationException e) {
+                throw new PrestoException(COMPILER_ERROR, e);
+            }
+        };
+    }
+
+    private ClassDefinition defineProjectionClass(RowExpression projection, InputChannels inputChannels, CallSiteBinder callSiteBinder)
+    {
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName(PageProjection.class.getSimpleName()),
+                type(Object.class),
+                type(PageProjection.class));
+
+        FieldDefinition blockBuilderField = classDefinition.declareField(a(PRIVATE, FINAL), "blockBuilder", BlockBuilder.class);
+
+        CachedInstanceBinder cachedInstanceBinder = new CachedInstanceBinder(classDefinition, callSiteBinder);
+        generatePageProjectMethod(classDefinition, blockBuilderField);
+        generateProjectMethod(classDefinition, callSiteBinder, cachedInstanceBinder, projection, blockBuilderField);
+
+        // getType
+        BytecodeExpression type = invoke(callSiteBinder.bind(projection.getType(), Type.class), "type");
+        classDefinition.declareMethod(a(PUBLIC), "getType", type(Type.class))
+                .getBody()
+                .append(type.ret());
+
+        // isDeterministic
+        classDefinition.declareMethod(a(PUBLIC), "isDeterministic", type(boolean.class))
+                .getBody()
+                .append(constantBoolean(determinismEvaluator.isDeterministic(projection)).ret());
+
+        // getInputChannels
+        classDefinition.declareMethod(a(PUBLIC), "getInputChannels", type(InputChannels.class))
+                .getBody()
+                .append(invoke(callSiteBinder.bind(inputChannels, InputChannels.class), "getInputChannels").ret());
+
+        // toString
+        String toStringResult = toStringHelper(classDefinition.getType()
+                .getJavaClassName())
+                .add("projection", projection)
+                .toString();
+        classDefinition.declareMethod(a(PUBLIC), "toString", type(String.class))
+                .getBody()
+                // bind constant via invokedynamic to avoid constant pool issues due to large strings
+                .append(invoke(callSiteBinder.bind(toStringResult, String.class), "toString").ret());
+
+        // constructor
+        generateConstructor(classDefinition, cachedInstanceBinder, method -> {
+            Variable thisVariable = method.getThis();
+            BytecodeBlock body = method.getBody();
+            body.append(thisVariable.setField(
+                    blockBuilderField,
+                    type.invoke("createBlockBuilder", BlockBuilder.class, newInstance(BlockBuilderStatus.class), constantInt(1))));
```

Comment:
With the change to make `reset` lazy, we only allocate the new memory one the block builder is used, so the page processor isn't holding on to memory.

As for the blocks created from the page processor it is a bit more complex.  The overallocation is that you mention above is a problem for top level columns that have a fixed width data types, but for variable with blocks we need extra buffer since the output size is not known ahead of time.  Additionally, many nested data structures are variable width so even fixed  width data types are not know ahead of time.  If we decide this is an issue we want to fix, we would need another reset method that takes expected position count, and I'm not sure the extra complexity is worth it at this state.